### ISSUE TITLE
cached augmented diffs differ from on-demand ones

### ISSUE THREAD

**Issue by tyrasd** [drolbr/Overpass-API#346]
Yesterday I was downloading some "live" augmented diffs (that are now served from a cache, see #342), but got one empty file (`2229363.osc`). After downloading the same (now, non cached) adiff from http://overpass-api.de/api/augmented_diff?id=2229363, I do actually get some data. Looking around the downloaded files a bit more and comparing them to their currently generated counterparts, I see that most (all?) of them actually don't match up with each other (the current adiffs containing more data then the disk-cached adiffs).

Here's an example: https://gist.github.com/tyrasd/fe213e1d17fc95c3a43b669b1ef154e1

**Comment by mmd-osm** [comment]
Can you try a cross check with http://dev.overpass-api.de/api_mmd/augmented_diff_status - it doesn't have caching enabled yet, i.e. it might be easier for comparison, also debug=true still works: http://dev.overpass-api.de/api_mmd/augmented_diff?debug=true&id=2230741

**Comment by tyrasd** [comment]
yeah, I just tried it with adiff 2230746. Here's the results in chronological order:

* the api_mmd instance (canonical url http://dev.overpass-api.de/api_mmd/augmented_diff?id=2230746) first returned me a *5.4MB* file (with `osm_base="2016-12-09T10:03:02Z"`)
* the main instance (cached adiff via canonical url http://overpass-api.de/api/augmented_diff?id=2230746) returns a *422kB* file (with `osm_base="2016-12-09T10:02:02Z"`)
* the main instance (manual query: http://overpass-turbo.eu/s/kxI) returns a *5.5MB* file (with `osm_base="2016-12-09T10:04:02Z"`)
* a second try a few minutes later on the api_mmd instance (canonical url http://dev.overpass-api.de/api_mmd/augmented_diff?id=2230746) returns a *5.5MB* file (with `osm_base="2016-12-09T10:08:02Z"`)

**Comment by tyrasd** [comment]
Looking at the actual (OSM) minutely-diffs, it seems like not all of the updates for each minute is contained in the same diff: [220430](http://planet.openstreetmap.org/replication/minute/002/220/430.osc.gz) contains data until [`2016-12-09T10:02:02Z`](http://planet.openstreetmap.org/replication/minute/002/220/430.state.txt), but the following two minutely diffs ([220431](http://planet.openstreetmap.org/replication/minute/002/220/431.osc.gz) and [220432](http://planet.openstreetmap.org/replication/minute/002/220/432.osc.gz)) contain updates for the `2016-12-09T10:01` minute as well. :confused: 

Is that to be expected? //cc @tomhughes

How could Overpass circumvent this issue? Waiting a few minutes before actually publishing an augmented diff maybe? (But then, how long would such a buffer period have to be?)

**Comment by tomhughes** [comment]
You can't assume time ordering of the diffs, because changes only appear in a diff once the relevant database transaction has committed so a long transaction may start before a short one but finish after it meaning the changes from the short one appear in an earlier diff.

**Comment by mmd-osm** [comment]
@tomhughes : I'm wondering if a transaction's creation timestamp is set right when the transaction starts, rather than immediately before the commit? Is there maybe some rule of thumb how long such a transaction could take in the worst case?

I assume that a transaction starts with "POST /api/0.6/changeset/#id/upload" and the commit is only triggered right after the call. If that's the case a very slow upload might indeed introduce quite some delay.

Not sure what the best way to deal with the issue would be.

**Comment by tomhughes** [comment]
You're confusing database transactions and OSM changesets. They are not the same thing.

Creating a changeset is a transaction. Uploading a single object to it is a transaction. Uploading a diff to it is another transaction. Closing it (if done explicitly) is yet another transaction.

The only thing you need to know is that you can't assume changes in the diffs will appear to be time ordered because they won't be.

I think most timestamps are assigned by ruby, not by the database, so for changeset creation will probably actually be before the transaction is even opened, but will only be visible in the diffs once the transaction closes. Same for objects in a diff - each one will be assigned as it is read by rails I think, but check the source if you want to be sure.

**Comment by tyrasd** [comment]
[Here](https://lists.openstreetmap.org/pipermail/dev/2016-August/029426.html) @lonvia mentioned that object timestamps should be set by the db

**Comment by mmd-osm** [comment]
> You're confusing database transactions and OSM changesets.

Well, I thought I was talking about HTTP calls being atomic and representing one transaction each. So, yes, there may be multiple db transactions associated to a single OSM changeset.

I think the main issue here is, that we have quite some semantic differences:
* a diff file represents a 'transport' container of whatever data is ready to be replicated at a given point in time. In case a transaction is still ongoing, it can be easily included in a later diff file. In particular, there's no guarantee that every object with a certain create/modification/deletion timestamp eventually ends up in a diff file, which is calculated based on the object's timestamp! That was one of the underlying assumptions, which turns to be incorrect.
* An augmented diff is supposed to include all objects with create/change/deletion timestamp within a certain 1 minute timeframe. As soon as such an augmented diff is available, it has - by definition - to include all objects, as it matches exactly a 1 minute timeframe, based on the object timestamp. Delivering that object "later", as with minutely diffs, does not match that concept.

Indeed very tricky question...

**Comment by tyrasd** [comment]
> Indeed very tricky question...

The only real solution I can currently think of would be for Overpass to store an additional meta field for each object version, namely the minutely diff number (or timestamp) in which it was published. Then, the augmented diff call can produce output in 1:1 correspondence to OSM's minutely diffs, i.e. one augmented diff for each OSM minutely diff including the same objects. That would be incompatible to Overpass' current augmented diff API, though.

@mmd-osm in #342 you mentioned that *several clients are [currently requesting] augmented diffs*. Do you know which applications those are? How did they not notice that a significant portion of the data is missing in the augmented diffs (if consumed immediately)?

**Comment by mmd-osm** [comment]
@tyrasd :  Regarding the clients: there are some AWS clients without User Agent, and @pa5cal / @geonick experimenting with it. So I'm not sure which applications are really using it atm.

If possible, changing the meta data for this use case, should probably be avoided. Maybe we have some alternative: 

Each minutely diff comes with a state.txt file, which according to [osmosis](https://github.com/openstreetmap/osmosis/blob/master/osmosis-apidb/src/main/java/org/openstreetmap/osmosis/apidb/v0_6/impl/ReplicationState.java) has the following contents.

Field | Description
------------ | -------------
 txnMax |The maximum transaction id in the database.
txnMaxQueried|The maximum transaction id currently replicated from the database.
 txnActive|The currently active transaction ids.
txnReady|The previously active transaction ids that can now be queried.
timestamp|The maximum timestamp of data currently read from the database.
sequenceNumber|The replication sequence number.


Important point here is, that `timestamp` just represents the maximum timestamp of the data. It does _not_ indicate, that all data up to and including this timestamp are included in that particular minutely diff.

I was hoping that we could leverage the information in txnActive and track, how long transactions are running. Assumption would be:
* If `txnActive` is empty, no further data is to be expected for timeframe up to `timestamp`
* If `txnActive` is not empty in minutely diff _0.state.txt_, we need to track those transaction numbers in subsequent state files (_1.state.txt_, _2.state.txt_, ...) , and only if all of the activate transaction numbers are no longer listed there, all data according to the timestamp in _0.state.txt_ should be available now.

@tomhughes : I didn't find much documentation on the state.txt semantics yet. Does such an `txnActive` based approach make some sense to you?

**Comment by tyrasd** [comment]
> `txnActive` is not empty […]

Oh, cool. Nice idea! I didn't know about those fields.

**Comment by mmd-osm** [comment]
Based on all minutely diff state files, I tried to calculate a distribution for transaction length according to `txnActiveList`: [distribution.txt](https://github.com/drolbr/Overpass-API/files/643946/distribution.txt). First column indicates the frequency, the second column the number of minutely diff files a certain transaction number appears (~ transaction length).

The good news is that a large amount of transactions are active for at most a few minutes only:
-  1 minute : 94%
- <= 10 minutes: 99.82%
- <= 60 minutes: 99.98%

But there are also some extreme outliers, taking up to 8 days (!), like transaction number 260199054 below. Not sure that happened there...
 

Transaction length in minutes  / transaction number:

```
  11516 260199054
  11480 260215678
  11456 260231443
  11250 260410000
  11239 260414562
   9712 262515052
   6907 228809845
   6860 228846931
   6850 228854287
   6765 228898323
   6633 228987460
   6237 229189047
   6208 229197102
   6182 229203673
   6176 229204729
   6174 229205129
   6166 229206498
   6163 264016655
   6156 264019446
   6125 264031042
   6115 264037662
   6096 229220543
   6094 229220931
   6093 229221077
   6093 229221022
   6091 229221296
   6089 229221594
   5859 229263378
   5708 774776540
   5660 229336030
```

**Comment by tyrasd** [comment]
> Transaction length in minutes / transaction number:

Looking at those ids, it seems like they are all from the late 2012 period (somewhat shortly after the redaction period). These also seems to have happened in clumps: e.g. this is the last state file that  http://planet.osm.org/replication/minute/000/106/720.state.txt contains many of those long-"running" transactions with the 260* ids. http://planet.osm.org/replication/minute/000/106/721.osc.gz doesn't seem to contain any suspicious (e.g. >> 1 minute old) data at first glance.

> Downside of this approach - it only works with minutely diffs.

AFAIK, hourly/dialy diffs are just concatenated minutely diffs. So, the open transactions for a hourly diff would be the same as for the last minutely diff it was made from.

**Comment by mmd-osm** [comment]
> Looking at those ids, it seems like they are all from the late 2012 period

Right, I should have mentioned that they were the just the most expensive ones. As the list is quite lengthy, I put it up on the dev server now, as it just won't fit on Github: http://dev.overpass-api.de/tmp/psv/transactions.txt.gz 

Another period with long running transactions was in Feb 2016, where we had the minutely replication issue with 001/788/263.osc.gz. Transaction 774776540 was active back then for about 4 days.

```
#Tue Feb 16 22:18:05 UTC 2016
sequenceNumber=1794968
txnMaxQueried=774777279
txnReadyList=
timestamp=2016-02-16T22\:18\:03Z
txnMax=774777279
txnActiveList=774776540,774777277
```

> AFAIK, hourly/dialy diffs are just concatenated minutely diffs. So, the open transactions for a hourly diff would be the same as for the last minutely diff it was made from.

The issue is that I don't know exactly how that process works. The wiki states that those files are being generated a few minutes after the hour and they're aggregating minutely diffs.

**Comment by drolbr** [comment]
Let's collect the valid and invalid assumptions:

> The only thing you need to know is that you can't assume changes in the diffs will appear to be time ordered because they won't be.

This is a huge problem because this assumption is deeply built into Overpass API. In particular that the timestamp of the diff means that all changes up to that date are included.

> AFAIK, hourly/dialy diffs are just concatenated minutely diffs. So, the open transactions for a hourly diff would be the same as for the last minutely diff it was made from.

I don't think so. Some time ago the main API may have dropped intermediate versions of objects from hourly or daily diffs. Otherwise, there would not be no gain of hourly and daily diffs over a tar file of minute diffs.

To sum things up:

A short term solution could be to add a fixed some-minute delay. The right amount of minutes to lose very few changes should be figured out from mmds statistics.

@tomhughes  Is the txActive field (presence and precise semantics) part of the long term interface? I suppose not. Because that would take away degrees of freedom from the main DB server.

As I assume that txActive is not part of the long term interface, I'm reluctant to build dependencies to that. However, I'm grateful for the statistics to understand what is happening right now.

I prefer as a long time solution to revisit the changelog database file and store the file from which the change has come. That way we could augment whatever change file comes in.

**Comment by tyrasd** [comment]
>> AFAIK, hourly/dialy diffs are just concatenated minutely diffs. So, the open transactions for a hourly diff would be the same as for the last minutely diff it was made from.
> I don't think so. Some time ago the main API may have dropped intermediate versions of objects from hourly or daily diffs. Otherwise, there would not be no gain of hourly and daily diffs over a tar file of minute diffs.

Is that really so?? According to the counterexample [mentioned recently on @dev](https://lists.openstreetmap.org/pipermail/dev/2016-November/029600.html), daily diffs do still contain intermediate versions of objects.

**Comment by mmd-osm** [comment]
@tyrasd : that's right, I used the term "aggregated" above, but it's really only some kind of concatenation. All intermediate versions are still present in hourly and daily diffs.

**Comment by mmd-osm** [comment]
> @tomhughes Is the txnActiveList field (presence and precise semantics) part of the long term interface? I suppose not. Because that would take away useful degrees of freedom from the main DB server.
> 
> As I assume that txnActiveList is not part of the long term interface, I'm reluctant to build dependencies to that. However, I'm grateful for the statistics to understand what is happening right now.

@brettch: please correct me if I'm wrong. Even the replication logic in osmosis is pretty much tight to the Postgresql transaction snapshot feature, see https://github.com/openstreetmap/osmosis/blob/master/osmosis-apidb/src/main/java/org/openstreetmap/osmosis/apidb/v0_6/impl/TransactionDao.java#L42 and https://www.postgresql.org/docs/current/static/functions-info.html#FUNCTIONS-TXID-SNAPSHOT.

In particular, the list of currently active transactions in state.txt comes directly from Postgresql's txid_current_snapshot() -> xip_list (=Active txids at the time of the snapshot.). 

This has been this way for several years now. Changing this process would have a large impact on osmosis as well.

**Comment by ltog** [comment]
Is the assumption correct, that there is no missing/doubled data when one consumes only cached adiffs, i.e. when you consume adiffs within 60 minutes after their existence was announced by `api/augmented_diff_status`?

(The 60 minutes stem from https://github.com/drolbr/Overpass-API/blob/ea9bb5521ecf36e8e66b5a0342e26459d6a2ffe3/src/bin/augmented_diff_loop.sh#L28 )

**Comment by brettch** [comment]
@mmd-osm, all your comments about how transaction processing works look right to me, but here is some additional context in case it's useful.  To be honest, it's been many years since I wrote most of this and my memory is getting a bit hazy ;-)

Osmosis replication processing ignores dates entirely and just uses the ```txid_current_snapshot()``` PostgreSQL function to obtain a snapshot of current database transaction state.

The minute diffs are the only ones that hit the DB directly.  The hour and day diffs just roll up the minute diffs into batches.  All diffs are a complete history of all changes between transaction points in the database and may contain multiple changes for a single entity.  In all cases they are sorted by nodes, ways, then relations, and in increasing id/version order (so hour and day diffs are NOT a simple concatenation of minute and hour diffs, they are re-sorted).  Hour and day diffs don't serve any purpose other than reducing the overhead of catching up over a long time interval.

None of the diffs are exactly time aligned, they are transaction snapshot aligned.  Minute diffs are simple scheduled to run once per minute and read all data that has landed since the last invocation.  If the planet server goes down for a while they'll end up quite large and contain more than a minute of data.  I think there's a maximum transaction limit per file in there somewhere to keep a lid on things.  Hour diffs feed off them and put approximately 1 hour of diffs into a file, but each minute file is placed entirely into a single hour file and never split across two files.

Some consequences of all this.  The data is aligned based on the timestamp of when the transaction was committed, not when the data was inserted.  Long running transactions may result in their data appearing in a later replication file than the entity timestamps would indicate.

Any tool that makes assumptions about the maximum duration of a transaction is doomed to occasional data loss.  Early Osmosis replication used timestamps.  It worked well in the early MySQL days, but was a disaster once the PostgreSQL DB and changesets were introduced.  Every time I tried increasing the "lag interval" to try to ensure all transactions had landed somebody invented a longer running transaction.

The content of the state files is not really meant for client tool consumption, only for Osmosis itself.  There wasn't a lot of thought into defining a public API.  You can use them as a source of diagnostic information (e.g. timestamp) but that's about it.  ```txnActiveList``` in particular is only useful for a tool hitting the database directly (i.e. Osmosis creating the minute files).  I probably should have put it elsewhere but it was simpler to keep it all in one place.

**Comment by paulmach** [comment]
Hi, I have been working though this problem for a while now as part of a side project to create accurate diffs and before/after states. The problem being, as described above, the timestamp of a object doesn't necessarily correspond to when it appeared in the database and the minutely diffs.  A lot of the minutely diffs have data from before the minute (usually a handful of seconds only) and even after since the generation doesn't seems to start until 2-3ish seconds after the minute.

The solution I've come up with is to compute a "committed at time" for each element. This is the estimated time the element's upload transaction completed.  The "augmented diff" becomes the planet minutely diff plus stuff that was committed before anything in that diff. So it's keyed off the minutely diff id, and not absolute time.

I compute the "committed at time" using the minutely diffs. An object's committed at time is the max timestamps for objects with matching changeset ids in the minutely diff. I think this works since the minutely diffs are a set of finished transaction data. It's not exactly correct since there could be two uploads for a given changeset in a given minute, but the short answer is I think it'll still work. :)  For elements added before minutely diffs started I do some estimates but it can't be completely correct.

I've been working on this as part of my own stack. I noticed the issues with overpass and didn't want to propose such a big change because I feel like most of the use cases for overpass wouldn't justify it.

There are many uses for being able to compute a true before/after state for any time and any area that I'm interested in exploring, but unfortunately I don't have anything to demo quite yet.

**Comment by tyrasd** [comment]
In http://www.openstreetmap.org/user/geohacker/diary/40846, it sounds like @geohacker found a workaround for this issue:

> A changeset being closed doesn't mean that all features that changed have been committed to the OSM database, and appear in the minutely diff right after. Some features may take longer to commit to the database, we handle these by updating the augmented diff from S3, and then recreating the changeset JSON.

@geohacker: How do you do this exactly?

**Comment by geohacker** [comment]
@tyrasd I was going to post here, this is a thorough hack but works alright. 

* I hooked a script on our Overpass instance to write minutely augmented diffs to an S3 location. 
* Then, we listen to the OSM minutely replication, prepare a list of all the feature timestamps that appear.
* If the required state is the current minute, download adiff from S3.
* If the state required isn't in Overpass we put the task back in a queue.
* If an old minutely adiff state is needed, request this via the Overpass API, and update adiff on S3 and update the changeset with new features. 

This relies entirely on the fact that Overpass adiff states are consistent and can be updated even if a feature arrives later in a minutely replication. And also that we host Overpass instance to entirely do this so it's a bit more reliable.

**Comment by geohacker** [comment]
Oh and the augmented diffs are also available for download. https://s3-ap-northeast-1.amazonaws.com/overpass-db-ap-northeast-1/augmented-diffs/

The state of the latest augmented diff is in a file called latest, like https://s3-ap-northeast-1.amazonaws.com/overpass-db-ap-northeast-1/augmented-diffs/latest.

You can request for an augmented diff this way: https://s3-ap-northeast-1.amazonaws.com/overpass-db-ap-northeast-1/augmented-diffs/2409184.osc

Hope this helps and also reduce load on the Overpass instance!

**Comment by mmd-osm** [comment]
> This relies entirely on the fact that Overpass adiff states are consistent and can be updated even if a feature arrives later in a minutely replication

As mentioned before in the blog comment, there's one drawback you need to be aware of: most data consumers would assume an augmented diff not to change once it has been published. For your use case, that's ok, because users will typically fetch data at a much later point in time for changeset analysis purposes.
In other cases (like updating some stats based on augmented diff), you might get screwed by those behind the scenes update of augmented diffs... That's basically how Overpass behaves right now, if you call the underlying query behind the augmetned_diff call across a number of hours.

**Comment by mmd-osm** [comment]
I implemented a very lightweight approach in https://github.com/mmd-osm/Overpass-API/wiki/Settings-for-0.7.58mmd-branch#other-files, which basically checks for an empty txnActiveList, and writes an additional file osm_base_completed_version.

Assumption here is that an empty txnActiveList implies that all pending transactions have been completed, and we should have all data for the given timestamp. 

https://github.com/mmd-osm/Overpass-API/wiki/Settings-for-0.7.58mmd-branch#rules_delta_completedsh provides an example on how to control area updates and make sure that we always have the complete data.

Same approach could be used for augmented diffs as well, by delaying creating new augmented diffs, until txnActiveList becomes empty. There may be some delays once in a while, still it's the easiest way as of today to ensure augmented diffs are complete. In addition, exactly zero changes are needed to the C++ backend, which is a big plus.
